crates/bevy_ecs/src/event/mut_iterators.rs

```diff
@@ -0,0 +1,275 @@
+        // Iterate the oldest first, then the newer events
+        let chain = a.iter_mut().chain(b.iter_mut());
+
+        Self {
+            mutator,
+            chain,
+            unread: unread_count,
+        }
+    }
+
+    /// Iterate over only the events.
+    pub fn without_id(self) -> EventMutatorIterator<'a, E> {
+        EventMutatorIterator { iter: self }
+    }
+}
+
+impl<'a, E: Event> Iterator for EventMutatorIteratorWithId<'a, E> {
+    type Item = (&'a mut E, EventId<E>);
+    fn next(&mut self) -> Option<Self::Item> {
+        match self
+            .chain
+            .next()
+            .map(|instance| (&mut instance.event, instance.event_id))
+        {
+            Some(item) => {
+                detailed_trace!("EventReader::iter() -> {}", item.1);
+                self.mutator.last_event_count += 1;
+                self.unread -= 1;
+                Some(item)
+            }
+            None => None,
+        }
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        self.chain.size_hint()
+    }
+
+    fn count(self) -> usize {
+        self.mutator.last_event_count += self.unread;
+        self.unread
+    }
+
+    fn last(self) -> Option<Self::Item>
+    where
+        Self: Sized,
+    {
+        let EventInstance { event_id, event } = self.chain.last()?;
+        self.mutator.last_event_count += self.unread;
+        Some((event, *event_id))
+    }
+
+    fn nth(&mut self, n: usize) -> Option<Self::Item> {
+        if let Some(EventInstance { event_id, event }) = self.chain.nth(n) {
+            self.mutator.last_event_count += n + 1;
+            self.unread -= n + 1;
+            Some((event, *event_id))
+        } else {
+            self.mutator.last_event_count += self.unread;
+            self.unread = 0;
+            None
+        }
+    }
+}
+
+impl<'a, E: Event> ExactSizeIterator for EventMutatorIteratorWithId<'a, E> {
+    fn len(&self) -> usize {
+        self.unread
+    }
+}
+
+/// A parallel iterator over `Event`s.
+#[derive(Debug)]
+#[cfg(feature = "multi_threaded")]
+pub struct EventMutatorParIter<'a, E: Event> {
+    mutator: &'a mut ManualEventMutator<E>,
+    slices: [&'a mut [EventInstance<E>]; 2],
+    batching_strategy: BatchingStrategy,
+    unread: usize,
+}
+
+#[cfg(feature = "multi_threaded")]
+impl<'a, E: Event> EventMutatorParIter<'a, E> {
+    /// Creates a new parallel iterator over `events` that have not yet been seen by `mutator`.
+    pub fn new(mutator: &'a mut ManualEventMutator<E>, events: &'a mut Events<E>) -> Self {
+        let a_index = mutator
+            .last_event_count
+            .saturating_sub(events.events_a.start_event_count);
+        let b_index = mutator
+            .last_event_count
+            .saturating_sub(events.events_b.start_event_count);
+        let a = events.events_a.get_mut(a_index..).unwrap_or_default();
+        let b = events.events_b.get_mut(b_index..).unwrap_or_default();
+
+        let unread_count = a.len() + b.len();
```

Comment:
the `debug_assert_eq!(unread_count, reader.len(events))` was removed here as well. was this intentional?